Google.Api.Gax.Grpc/Rest/PartialDecodingStreamReader.cs

```diff
@@ -115,32 +117,44 @@ public PartialDecodingStreamReader(Task<TextReader> textReaderTask, Func<string,
     /// <inheritdoc />
     public async Task<bool> MoveNext(CancellationToken cancellationToken)
     {
-        while (true)
+        bool disposeOnExit = true;
+        try
```

Comment:
I think that if you dispose in the catch and rethrow, you don't need this variable. _endOfData would have already disposed on ReadAndProcessData.

Or, stop disposing on ReadAndProcessData and only dispose on MoveNext...

Or, if you check for cancellation on ReadAndProcessData only (I know you prefer checking always on MoveNext) then you can dispose and throw on ReadAndProcessData only.

Reply:
I'm trying to avoid catching and rethrowing, basically - that always feels a bit weird.
Also, in the new PR I expect to separate out disposal part from everything else (at the same time as changing the exceptions being thrown). I think it'll be simpler at that point, when it's obvious that the "dispose on the way out" doesn't affect the result itself at all.